### PR TITLE
Changes to add NCAR's Casper/DAV systems to the machine lists

### PR DESCRIPTION
**Pull request by gdicker1**
This setup defaults to adding "-acc -ta=tesla:cc70" to the build flags
with PGI compiler and utilizing the NTDD0002 reservation for jobs
submitted on Casper.

[ Description of the changes in this Pull Request. It should be enough
information for someone not following this development to understand.
Lines should be wrapped at about 72 characters. Please also update
the CIME documentation, if necessary, in doc/source/rst and indicate
below if you need to have the gh-pages html regenerated.]

Test suite: 
Test baseline: 
Test namelist changes: 
Test status: [bit for bit, roundoff, climate changing]

Fixes [CIME Github issue #]

User interface changes?: No

Update gh-pages html (Y/N)?: N

Code review:

### CODE REVIEW

File: config/cesm/machines/config_machines.xml [hunk 1]
```diff
@@ -352,6 +352,90 @@ This allows using a different mpirun command to launch unit tests
     </resource_limits>
   </machine>
 
+   <machine MACH="casper">
+    <DESC>NCAR GPU platform, os is Linux, 32 pes/node, batch system is slurm</DESC>
+    <NODENAME_REGEX>casper*</NODENAME_REGEX>
+    <!-- MPT sometimes timesout at model start time, the next two lines cause
+    case_run.py to detect the timeout and retry FORCE_SPARE_NODES times -->
+    <MPIRUN_RETRY_REGEX>openmpi: Launcher network accept (MPI_LAUNCH_TIMEOUT) timed out</MPIRUN_RETRY_REGEX>
```
Comment (jedwards4b) [inline config/cesm/machines/config_machines.xml:360]: Have you tested this?  This code is mpt specific along with several other settings here - can you remove them? (The mpt specific stuff).

File: config/cesm/machines/config_machines.xml [hunk 2]
```diff
@@ -352,6 +352,90 @@ This allows using a different mpirun command to launch unit tests
...
+      </modules>
+      <modules>
+        <command name="load">ncarcompilers/0.5.0</command>
+      </modules>
+    </module_system>
+    <environment_variables>
+      <env name="MODULEPATH">/glade/u/apps/dav/modulefiles/default/compilers:/glade/u/apps/dav/modulefiles/default/idep</env>
+      <env name="OMP_STACKSIZE">256M</env>
+      <env name="TMPDIR">/glade/scratch/$USER</env>
+      <env name="MPI_TYPE_DEPTH">16</env>
```
Comment (jedwards4b) [inline config/cesm/machines/config_machines.xml:421]: mpt specific

File: config/cesm/machines/config_machines.xml [hunk 3]
```diff
@@ -352,6 +352,90 @@ This allows using a different mpirun command to launch unit tests
...
+      <env name="CESMDATAROOT">/glade/p/cesmdata/cseg</env>
+      <env name="NETCDF_PATH">/glade/u/apps/dav/opt/netcdf-mpi/4.7.3/openmpi/3.1.4/pgi/19.9</env>
+      <env name="MPI_IB_CONGESTED">1</env>
```
Comment (jedwards4b) [inline config/cesm/machines/config_machines.xml:424]: MPT specific

File: config/cesm/machines/config_machines.xml [hunk 4]
```diff
@@ -352,6 +352,90 @@ This allows using a different mpirun command to launch unit tests
...
+      <env name="MPI_USE_ARRAY"/>
```
Comment (jedwards4b) [inline config/cesm/machines/config_machines.xml:425]: mpt specific

File: config/cesm/machines/config_machines.xml [hunk 5]
```diff
@@ -352,6 +352,90 @@ This allows using a different mpirun command to launch unit tests
...
+    </environment_variables>
+    <environment_variables unit_testing="true">
+      <env name="MPI_USE_ARRAY">false</env>
+    </environment_variables>
+    <environment_variables queue="share">
+      <env name="TMPDIR">/glade/scratch/$USER</env>
```
Comment (jedwards4b) [inline config/cesm/machines/config_machines.xml:431]: not needed

File: config/cesm/machines/config_machines.xml [hunk 6]
```diff
@@ -352,6 +352,90 @@ This allows using a different mpirun command to launch unit tests
...
+      <env name="MPI_USE_ARRAY">false</env>
```
Comment (jedwards4b) [inline config/cesm/machines/config_machines.xml:432]: mpt specific

File: config/cesm/machines/config_machines.xml (hunk 1 above)
Comment (gdicker1) [reply config/cesm/machines/config_machines.xml:360]: This change is part of a porting of MOM6 using OpenACC by NCAR. We have checked that this allows for the code to run on our single node on Casper, and we are currently in the process of verifying the results with the PGI compiler on Cheyenne.
We have not run the regression test suite with these changes.

**Comment by jedwards4b** [comment]
I'm not asking you to run the scripts_regression_tests.py - although you should.  I am merely pointing out that you cut and pasted a lot of configuration information that is specific to the mpt library on cheyenne and which is not relevant for this port.  I've tried to point out where, please remove it.

**Comment by gdicker1** [comment]
Good to know we're on the same page. I'm double checking parts of this with my colleagues after lunch and should have an update soon. Thank you for pointing out the MPT specific parts.

**Comment by gdicker1** [comment]
Could you double check me on the retry regex? I'm not sure if that's MPT specific or not (and if my edit even does anything). Other than that this currently works for the 1 node case on Casper.

File: config/cesm/machines/config_compilers.xml [hunk 7]
```diff
@@ -653,6 +653,23 @@ using a fortran linker.
   </SLIBS>
 </compiler>
 
+<compiler MACH="casper" COMPILER="pgi">
+  <FFLAGS>
+    <append> -Mnofma -acc -ta=tesla:cc70 -Minfo=accel </append>
+    <append> -I$(EXEROOT)/ocn/obj/FMS </append>
```
[the remark this answers is not in the document]
Comment (gdicker1) [reply config/cesm/machines/config_compilers.xml:659]: This is needed for a successful build of the CESM executable with PGI. If it is only included with LDFLAGS, compilation fails due to not being able to find an FMS related module.

File: config/cesm/machines/config_machines.xml [hunk 8]
```diff
@@ -352,6 +352,79 @@ This allows using a different mpirun command to launch unit tests
...
+   <machine MACH="casper">
+    <DESC>NCAR GPU platform, os is Linux, 36 pes/node, batch system is slurm</DESC>
+    <NODENAME_REGEX>casper*</NODENAME_REGEX>
+    <!-- MPT sometimes timesout at model start time, the next two lines cause
+    case_run.py to detect the timeout and retry FORCE_SPARE_NODES times -->
+    <MPIRUN_RETRY_REGEX>openmpi: Launcher network accept (MPI_LAUNCH_TIMEOUT) timed out</MPIRUN_RETRY_REGEX>
+    <MPIRUN_RETRY_COUNT>10</MPIRUN_RETRY_COUNT>
+    <OS>LINUX</OS>
+    <COMPILERS>pgi</COMPILERS>
+    <MPILIBS compiler="pgi" >openmpi</MPILIBS>
```
[the remark this answers is not in the document]
Comment (gdicker1) [reply config/cesm/machines/config_machines.xml:364]: I think we would like to leave this. We will most likely be adding Intel configurations as well at a later date to compare against.

File: config/cesm/machines/config_machines.xml [hunk 9]
```diff
@@ -352,6 +352,79 @@ This allows using a different mpirun command to launch unit tests
...
+      <modules>
+        <command name="load">ncarcompilers/0.5.0</command>
+      </modules>
+    </module_system>
+    <environment_variables>
+      <env name="MODULEPATH">/glade/u/apps/dav/modulefiles/default/compilers:/glade/u/apps/dav/modulefiles/default/idep</env>
+      <env name="OMP_STACKSIZE">256M</env>
+      <env name="TMPDIR">/glade/scratch/$USER</env> 
+      <env name="CESMDATAROOT">/glade/p/cesmdata/cseg</env>
+      <env name="NETCDF_PATH">/glade/u/apps/dav/opt/netcdf-mpi/4.7.3/openmpi/3.1.4/pgi/20.4</env>
```
[the remark this answers is not in the document]
Comment (gdicker1) [reply config/cesm/machines/config_machines.xml:421]: When I make this change and create a new case, upon running `case.setup` I get `ERROR: Not safe to leave unresolved items in env var value: '$NETCDF'`.

[the remark this answers is not in the document]
Comment (jedwards4b) [reply config/cesm/machines/config_machines.xml:421]: Sorry, should be $ENV{NETCDF}

File: config/cesm/machines/config_compilers.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (jedwards4b) [reply config/cesm/machines/config_compilers.xml:659]: This is the wrong location for this  - I believe that it should be in the buildlib for mom.

[the remark this answers is not in the document]
Comment (alperaltuntas) [reply config/cesm/machines/config_compilers.xml:659]: @gdicker1 , you can add the above include flag to the following line in MOM_interface as an additional include directory: https://github.com/ESCOMP/MOM_interface/blob/90bff5d3e8c41adc9768f3199948dba8171fb640/cime_config/buildlib#L66

**Comment by gdicker1** [comment]
Sorry, yes this should have new changes in the most recent push.

File: config/cesm/machines/config_compilers.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (gdicker1) [reply config/cesm/machines/config_compilers.xml:659]: I double checked this and the MOM6 buildlib does already have the FMS include for the MOM6 build. If FFLAGS isn't updated this way the step that builds the CESM executable fails. Adding this to LD_FLAGS doesn't solve the issue, but adding it to FFLAGS does.

**Pull request merged**